nat-lab: Refactor event handling

lcruz99 opened the pull request:
### Description
Some nat-lab tests were not handling events correctly, causing flakyness. After initial investigation on test_direct_connection_endpoint_gone() flakyness it was found that most calls to wait_for_event functions were checking historic events where they should instead be waiting for new events/states (node & derp servers). Also it was using wait_for_any_derp_state which looks for a state in any derp, where they should be checking on every derp (eg: blocking every derp servers connection).
In this PR, the checks were splitted into two distint category of functions, `wait_for_state_<derp/peer>` and `wait_for_event_<derp/peer>`. `wait_for_state_<derp/peer>` functions wait for a state report that matches with the given arguments, and `wait_for_event_<derp/peer>` functions wait for a new event report on stdout that matches the given states list. The difference is that the state functions checks (can also wait) for that state, while the event functions exclusively wait for new events from the instant they are called ~~OR they can take a asyncio.Event() as argument which is later set() in a way that the caller is able to launch the function before executing a specific action, ensuring that events are caught and thus avoiding data races~~ (Removed).

Note1: critical changes were introduced in telio.py and the rest are adjustments.
Note2: to avoid bigger MR, I decided to tackle [LLT-3880](https://nordsec.atlassian.net/browse/LLT-3880)
and [LLT-3846](https://nordsec.atlassian.net/browse/LLT-3846) on a separated MR.
Note3: probably a lock could solve the need for wait_for_event functions, it would synchronize quick event changes (Disconnected -> Connecting -> Connected). 

### :ballot_box_with_check: Definition of Done checklist
- [x] Commit history is clean
- [x] README.md is updated
- [x] changelog.md is updated

gytsto commented on nat-lab/tests/test_network_switch.py:
```diff
@@ -332,33 +328,33 @@ async def test_mesh_network_switch_direct(
         await alpha_client.notify_network_change()
 
         await testing.wait_lengthy(
-            alpha_client.wait_for_any_derp_state(
```
I see this doesn't work? waiting for all connection states I mean...

What if you were to write something like this:
```
async with run_async_contexts(
    [
        alpha_client.wait_for_event_on_any_derp([telio.State.Disconnected]),
        alpha_client.wait_for_event_on_any_derp([telio.State.Connecting]),
        alpha_client.wait_for_event_on_any_derp([telio.State.Connected]),
    ]
) as futures:
    # wait for futures to schedule
    await asyncio.sleep(0)
    await alpha_client.notify_network_change()
    for fut in futures:
        await fut
```

lcruz99 replied:
that would work I think.. is it worth to check those intermediate events when we only care about [telio.State.Connected]? 


I have the same question for other tests: if we want to wait for a node event with Direct path is it worth to check [telio.State.Connected] Derp event and [telio.State.Connected] node event? IMO we should have those intermediate steps on a just one test case to test specifically that, on all other tests it looks redundant.

gytsto replied:
I'd say it is important to track all the events, even if they are not really relevant to test. This way we are making sure, that correct flow is executed. But maybe let's leave this for the future.

And about direct path. It is important to check for derp first, since w/o derp, there cannot be any direct connection.

lcruz99 replied:
> And about direct path. It is important to check for derp first, since w/o derp, there cannot be any direct connection.
That's the point,

lcruz99 replied:
My though is to simplify tests as much as possible, to have transparent tests so that we easily understand them: adding common setup/destroy methods, cleaning test redundancy (example above), limit tests to one behaviour per test case, etc.

gytsto replied:
I think we will do this in the future, with nat-lab framework improvements. For now this PR is not about this I guess

lcruz99 replied:
correct 😄

gytsto commented on nat-lab/tests/test_node_state_flickering.py:
```diff
@@ -84,34 +84,28 @@ async def test_node_state_flickering(
-                    client_alpha.wait_for_any_new_derp_state(),
-                    client_beta.wait_for_any_new_node_event(alpha.public_key),
-                    client_beta.wait_for_any_new_derp_state(),
+                    client_alpha.wait_for_any_state_peer(beta.public_key),
```
events here, not states

lcruz99 replied:
good catch

gytsto reviewed (changes requested):
General issues:
1. Split derp / peer / conn _tracker events in separte gathers for better indication (also, `asyncio.gather` doesn't ensure you, that coroutines will be called in order you provided)
2. It feels like we have too much `wait_for_derp_(state/event)` functions, kinda getting confusing
3. providing `asyncio.Event()` in those wait functions not needed.
4. also need to thoroughly check tests if logic was not changed accidentally

gytsto commented on nat-lab/tests/telio.py:
```diff
@@ -213,57 +213,76 @@ async def notify_peer_state(
             await asyncio.sleep(0.01)
 
     def get_peer_info(self, public_key: str) -> Optional[PeerInfo]:
         return self._peer_info.get(public_key)
```
I think this can be changed to, therefore we wouldn't need `self._peer_info`
```
def get_peer_info(self, public_key: str) -> Optional[PeerInfo]:
    return [
        peer_event
        for peer_event in self._peer_state_events
        if peer_event.public_key == public_key
    ][-1]
```

lcruz99 replied:
Changed

lcruz99 commented on nat-lab/tests/telio.py:
```diff
@@ -213,57 +213,76 @@ async def notify_peer_state(
         old_events = _get_events()
+
+        if event:
+            event.set()
```
I think you missed the point. Event is there to synchronize tasks, after we initialize old_events the event is set() to inform the main thread that the event waiting task is ready (we got the historic events, no dataraces will happen), giving green light to execute the test actions. Therefore, moving it to While loop is trivial and adds unnecessary cpu iterations. 

You can do `future = run_async(notifiy_peer_event(...))`,  however how do you ensure that notifiy_peer_event() have started and old_events been fetched before starting test actions? You could await on that future but that would block. So I came up with Event() because it truly launchs the task asynchrounisly and confirms we are ready to move on.

lcruz99 replied:
Just to add that, we do not want to wait for the function return, but for the function signal that the old events have been fetched, allowing us to start the test actions without eventual dataraces.

gytsto replied:
Can you show some use cases of it, cuz I cannot wrap my head around, why is it actually needed and why it is not used anywhere. :confused:

lcruz99 replied:
Oops, it seems I did some last minute changes that I didn't remember, sorry about that, you are right.

lcruz99 replied:
So currently we are not using them, I'll have them removed. 

The thought behind it was to avoid dataraces. For instance, if we have an event that we would like to catch but it was logged so quickly after the test action that we couldn't call wait_for_event() in time, missing it.

lcruz99 replied:
It technically still can happen, see: https://github.com/NordSecurity/libtelio/pull/41/files#diff-778fbedd2d660ca2c2322076ce67ff83072b808e239308da0bbfc3f130add155R331

lcruz99 commented on nat-lab/tests/telio.py:
```diff
@@ -213,57 +213,76 @@ async def notify_peer_state(
         old_events = _get_events()
+
+        if event:
+            event.set()
```
Replied above

lcruz99 replied:
will remove

lcruz99 commented on nat-lab/tests/telio.py:
```diff
@@ -332,29 +351,56 @@ async def message_done(self, message_idx: int) -> None:
 
-    async def wait_for_derp_state(self, server_ip: str, states: List[State]) -> None:
-        await self._runtime.notify_derp_state(server_ip, states, False)
+    async def wait_for_any_state_peer(
```
removed

lcruz99 commented on nat-lab/tests/test_direct_connection.py:
```diff
@@ -80,13 +80,19 @@
 @pytest.mark.parametrize(
-    "endpoint_providers, client1_type, client2_type, reflexive_ip",
-    UHP_conn_client_types,
+    "endpoint_providers, client1_type, client2_type",
```
test leftovers, reverted

lcruz99 commented on nat-lab/tests/telio.py:
```diff
@@ -447,67 +493,120 @@ async def simple_start(self):
-        await self.get_events().wait_for_peer_state(
-            public_key, [State.Disconnected], [path]
-        )
+    async def wait_for_any_state_peer(
```
removed

lcruz99 commented on nat-lab/tests/telio.py:
```diff
@@ -447,67 +493,120 @@ async def simple_start(self):
+                await asyncio.sleep(0.01)
 
-    async def wait_for_new_derp_state(
+    async def wait_for_any_state_on_any_derp(
```
removed

lcruz99 commented:
> General issues:
> 
>     1. Split derp / peer / conn _tracker events in separte gathers for better indication (also, `asyncio.gather` doesn't ensure you, that coroutines will be called in order you provided)
> 
>     2. It feels like we have too much `wait_for_derp_(state/event)` functions, kinda getting confusing
> 
>     3. providing `asyncio.Event()` in those wait functions not needed.
> 
>     4. also need to thoroughly check tests if logic was not changed accidentally

1. Splitted some coroutings to 2 futures (asyncio.gather)  for better readability. You're right however coroutines inside gather() were not order dependent
2. Removed specific cases of `wait_for_derp_(state/event)` functions
3. Removed `asyncio.Event()` `wait_for_derp_(state/event)` functions
4. Hope everything's fine :)

gytsto commented on nat-lab/tests/telio.py:
```diff
@@ -166,7 +166,6 @@ class AdapterType(Enum):
 
 class Runtime:
     _output_notifier: OutputNotifier
-    _peer_info: Dict[str, PeerInfo]
```
simillar logic should apply to `_derp_state` , should be removed and used last event from `_derp_state_events`

lcruz99 replied:
missed that, fixed!

lcruz99 replied:
however, shouldn't we make `_derp_state_events/_peer_state_events` atomic? At the moment we have different threads writing/reading them without synchronization

gytsto replied:
Idk if we need it for now, but you could create Jira issue and add relation to LLT-4054

lcruz99 replied:
After some thinking, I've concluded that's not worth to make them atomic. We have only one thread writing them and even if race conditions happen it's not critical as the readers (wait_for_events) will just read them again with no side effects (notify state/event functions)

gytsto commented on nat-lab/tests/telio_test.py:
```diff
@@ -67,247 +65,98 @@ async def test_set_derp_state(self) -> None:
             'event node: "{"identifier":"tcli","public_key":"AAA","state":"connected","is_exit":true,"is_vpn":true,"ip_addresses":[],"allowed_ips":[],"endpoint":null,"hostname":null,"allow_incoming_connections":false,"allow_peer_send_files":false,"path":"relay"}"'
         )
-        peer_info = runtime.get_peer_info("AAA")
-        assert peer_info is not None and peer_info.state == State.Connected
```
nvm actually

gytsto commented on nat-lab/tests/telio_test.py:
```diff
@@ -328,43 +177,14 @@ async def test_explicit_new_derp_state(self) -> None:
-            )
-
-        # Start waiting for new event before it is being generated
-        async with run_async_context(
```
why remove this?

lcruz99 replied:
Removed this test initially because it didn't make sense for states, nevertheless added new tests

gytsto commented:
- I see you missed alot of comments :smile: 
- `telio_test.py` is missing tests for `notify_peer_event` and `notify_derp_event`
- also commit history should be cleaned up

I hope I didn't forget anything else :laughing: but let's wait for you to battle current comments first

lcruz99 commented:
> * I see you missed alot of comments 😄

still learning Github, it seems the comments I missed only appeared in 'Files changed tab', not that intuitive..

>     * `telio_test.py` is missing tests for `notify_peer_event` and `notify_derp_event`

Added tests

>     * also commit history should be cleaned up

Will be cleaned before merging 😄

lcruz99 commented on nat-lab/tests/test_derp_connect.py:
```diff
@@ -241,9 +273,15 @@ async def test_derp_reconnect_3clients() -> None:
-        await testing.wait_lengthy(check_derp_connection(alpha_client, DERP1_IP, True))
-        await testing.wait_lengthy(check_derp_connection(beta_client, DERP2_IP, True))
-        await testing.wait_lengthy(check_derp_connection(gamma_client, DERP3_IP, True))
+        await testing.wait_lengthy(
```
Added

lcruz99 commented on nat-lab/tests/test_derp_connect.py:
```diff
@@ -209,9 +230,15 @@ async def test_derp_reconnect_3clients() -> None:
-        await testing.wait_lengthy(check_derp_connection(alpha_client, DERP1_IP, True))
-        await testing.wait_lengthy(check_derp_connection(beta_client, DERP2_IP, True))
-        await testing.wait_lengthy(check_derp_connection(gamma_client, DERP2_IP, True))
+        await testing.wait_lengthy(
```
Added

lcruz99 commented on nat-lab/tests/test_derp_connect.py:
```diff
@@ -45,41 +47,49 @@ async def test_derp_reconnect_2clients() -> None:
         await exit_stack.enter_async_context(
-            beta_client.get_router().break_tcp_conn_to_host(DERP1_IP)
+            beta_client.get_router().break_tcp_conn_to_host(
+                str(DERP_SERVERS[0]["ipv4"])
```
dunno.. changed to `DERP1_IP`

lcruz99 commented on nat-lab/tests/test_derp_connect.py:
```diff
@@ -45,41 +47,49 @@ async def test_derp_reconnect_2clients() -> None:
+            asyncio.gather(
+                alpha_client.wait_for_state_derp(DERP1_IP, [State.Connected]),
+                beta_client.wait_for_state_derp(DERP1_IP, [State.Connected]),
+                alpha_client.wait_for_state_peer(beta.public_key, [State.Connected]),
```
done

lcruz99 commented on nat-lab/tests/test_connection_states.py:
```diff
@@ -85,24 +85,18 @@ async def test_connected_state_after_routing(
-        )
-
-        await testing.wait_long(
+        await testing.wait_lengthy(
```
you're right, splitted

lcruz99 commented on nat-lab/tests/test_connection_states.py:
```diff
@@ -17,32 +17,32 @@
-            telio.AdapterType.Default,
-            marks=pytest.mark.mac,
-        ),
+        # pytest.param(
```
yess, uncommented

arvydasv commented on nat-lab/tests/telio.py:
```diff
@@ -213,66 +209,88 @@ async def notify_peer_state(
-                )
-            ) or (new_event_explicitly and set(events) - set(old_events)):
-                self._peer_state_events = list(set(self._peer_state_events) - events)
+            new_events = _get_events()[len(old_events) :]
```
What does the square brackets do?

lcruz99 replied:
They assign to new_events a slice of the list returned by _get_events(). The slice is from len(old_events) index to the end of the list, which translates to every event except the old ones ones (on old_events).

arvydasv commented:
+1

gytsto reviewed (approved):
:+1: don't forget to update commit messages
